### ISSUE TITLE
shipping.php class, calculation of `shipping_num_boxes` can yield incorrect results

### ISSUE THREAD

**Issue by lat9** [zencart/zencart#5564]
While it's a kinda-sorta product/site misconfiguration, if a product is created that has a weight larger than identified in **Configuration :: Shipping/Packaging :: Enter the Maximum Package Weight you will ship** (`SHIPPING_MAX_WEIGHT`) and that single product is added to the cart, the class' calculation of `$shipping_num_boxes` indicates that multiple boxes are to be shipped!

To reproduce with the demo products:
1. Set that configuration value to 20
2. Navigate to the `/index.php?main_page=product_info&cPath=25&products_id=100` product, place it in the cart (it weighs 45 lbs).
3. Navigate to the shopping-cart page and request a shipping quote.

`$shipping_num_boxes` is calculated as 3 for that single item in the cart!

**Comment by scottcwilson** [comment]
This works great if you're shipping something that could easily be split in three parts.  Otherwise not so much. :(

**Comment by piloujp** [comment]
Would it not be better to filter this so you can't set the limit lower than the heaviest product in catalog and can't add a product heavier than the limit?
Anyway box number calculation is too simple and won't work well with products close to the max weight limit. For example 3 products weighting 16, 18 and 5 with a limit of 20 will give you 2 boxes (limit 20 each)...
I have some idea but I am not sure how to access/use necessary data like weight and quantity of each item in cart. Do we have to query database or are there other objects property/methods we can use when this class is created?

**Comment by scottcwilson** [comment]
> Would it not be better to filter this so you can't set the limit lower than the heaviest product in catalog and can't add a product heavier than the limit?

No, this would not be better.  Shipping methods which don't work with the cart contents should be disabled.

**Comment by lat9** [comment]
> This works great if you're shipping something that could easily be split in three parts. Otherwise not so much. :(

Of course, there's no way to indicate that a product can be 'split'.

**Comment by lat9** [comment]
> > Would it not be better to filter this so you can't set the limit lower than the heaviest product in catalog and can't add a product heavier than the limit?
> 
> No, this would not be better. Shipping methods which don't work with the cart contents should be disabled.

Er, shipping modules depend on a **_valid_** calculation of `$shipping_num_boxes` for their individual calculations.

**Comment by piloujp** [comment]
> Er, shipping modules depend on a valid calculation of $shipping_num_boxes for their individual calculations.

Yes it should if you want to use multiple boxes shipping.
When starting to make an adequate calculation algorithm I realized the all system has to be re-designed... In actual code 'multiboxing' is based on the assumption all shipping boxes rates are identical and at max weight limit! Only one shipping rate is calculated and multiplied by number of boxes.
Example:
Limit 20, two items, one at 18 and one at 3. Calculation will show you need two boxes at rate equivalent to weight 20.
Customer will definitely not be happy to be charged 20 rate for a 3 item.

**Comment by lat9** [comment]
Good point, @piloujp.  I believe that I've (again) opened a hornet's nest!

**Comment by scottcwilson** [comment]
The way it works now is a reasonable approximation that works for most stores.  
We should not change the entire shipping system - which is quite complicated - to deal with edge cases.  These people can buy Advanced Shipper or have a developer customize their code.

**Comment by lat9** [comment]
I would be happy if the base processing didn't chop products up into pieces in order to ship, as described in my initial posting.

**Comment by piloujp** [comment]
> I would be happy if the base processing didn't chop products up into pieces in order to ship, as described in my initial posting.

Your wish might be granted. I made a new function that produce an ordered (descending) array of items weights and quantities in the cart. Then comparing heaviest item (the first in the array) to max weight you can stop quoting. Here is the code:
```
public function get_weight_qty_contents()
    {
		global $db;
        $weight_quantity_array = array();
        if (is_array($_SESSION['cart']->contents)) {
            foreach ($_SESSION['cart']->contents as $products_id => $data) {
				$sql = "SELECT products_weight, product_is_always_free_shipping, products_virtual
                    FROM " . TABLE_PRODUCTS . "
                    WHERE products_id = " . (int)$products_id;
				if ($product = $db->Execute($sql)) {
					// adjusted count for free shipping
					if ($product->fields['product_is_always_free_shipping'] != 1 and $product->fields['products_virtual'] != 1) {
						$ind_product_weight = $product->fields['products_weight'];
					} else {
						$ind_product_weight = 0;
					}
				} else {
						$ind_product_weight = 0;
					}
				if (isset($weight_quantity_array[$ind_product_weight])) {
					$indqty = $weight_quantity_array[$ind_product_weight] + $_SESSION['cart']->contents[$products_id]['qty'];
				} else {
					$indqty = $_SESSION['cart']->contents[$products_id]['qty'];
				}
				$weight_quantity_array[$ind_product_weight] = $indqty;
            }
        }
		krsort($weight_quantity_array, SORT_NUMERIC);
        return $weight_quantity_array;
    }
```
Then in function quote just at the beginning after `global $shipping_weight, $uninsurable_value;` add this:
```
$heaviest_weight = array_keys($this->get_weight_qty_contents())[0];
		if ($heaviest_weight and $heaviest_weight >= SHIPPING_MAX_WEIGHT) {
            return;
		}
```
With quantities and weights together a proper calculation of number of boxes and their actual weight can be done two. Then a quote for each box could be kept in an array (new global variable) added to what quote function returns. Old shipping modules would not use it but new ones could take advantage of this.
What do you think of that? But first please have a look at the code above, it works but is just a draft and needs improvement.

**Comment by piloujp** [comment]
After testing this I added an algorithm for box number calculation that simply add weight from beginning of the array (heaviest product) until limit is reached then ignore next one after putting it in a new box (array) and add following one or not if limit is reach and again and again until end of array.
it works well and has the advantage of including a tare for each box and not one tare for total weigh only. It eliminates under estimation of shipping rate you don't have to cut products in half or more. It is just a modification on a small part of shipping.php file, not the whole shipping system.
Then I rewrite 3 of the basic shipping modules (perweightunit, table and zone) that use weight for shipping rate calculation so they can make a quote for each box and add them at the end.
Should I do a PR or not? Please tell me.

**Comment by scottcwilson** [comment]
> Should I do a PR or not?

Not.  You can build a plugin, provide customization instructions, write documentation, etc.

**Comment by piloujp** [comment]
OK, I will do that.
